cmd/query/app/server.go

```diff
@@ -193,37 +199,91 @@ type httpServer struct {
 
 var _ io.Closer = (*httpServer)(nil)
 
-func createHTTPServer(
-	ctx context.Context,
+func createHTTPRouter(
 	querySvc *querysvc.QueryService,
 	metricsQuerySvc querysvc.MetricsQueryService,
 	queryOpts *QueryOptions,
 	tm *tenancy.Manager,
 	telset telemetery.Setting,
-) (*httpServer, error) {
+) *mux.Router {
 	apiHandlerOptions := []HandlerOption{
 		HandlerOptions.Logger(telset.Logger),
 		HandlerOptions.Tracer(telset.TracerProvider),
 		HandlerOptions.MetricsQueryService(metricsQuerySvc),
 	}
-
 	apiHandler := NewAPIHandler(
 		querySvc,
 		tm,
 		apiHandlerOptions...)
+
 	r := NewRouter()
 	if queryOpts.BasePath != "/" {
 		r = r.PathPrefix(queryOpts.BasePath).Subrouter()
 	}
-
 	(&apiv3.HTTPGateway{
 		QueryService: querySvc,
 		TenancyMgr:   tm,
 		Logger:       telset.Logger,
 		Tracer:       telset.TracerProvider,
 	}).RegisterRoutes(r)
-
 	apiHandler.RegisterRoutes(r)
+
+	return r
+}
+
+func createHTTPServerOTEL(
+	ctx context.Context,
+	querySvc *querysvc.QueryService,
+	metricsQuerySvc querysvc.MetricsQueryService,
+	queryOpts *QueryOptions,
+	tm *tenancy.Manager,
+	telset telemetery.Setting,
+) (*httpServer, error) {
+	r := createHTTPRouter(querySvc, metricsQuerySvc, queryOpts, tm, telset)
+	var handler http.Handler = r
+	if queryOpts.BearerTokenPropagation {
+		handler = bearertoken.PropagationHandler(telset.Logger, handler)
+	}
+	recoveryHandler := recoveryhandler.NewRecoveryHandler(telset.Logger, true)
+	handler = recoveryHandler(handler)
+	s, err := queryOpts.HTTP.ToServer(
+		ctx,
+		telset.Host,
+		component.TelemetrySettings{
+			Logger:         telset.Logger,
+			TracerProvider: telset.TracerProvider,
+			LeveledMeterProvider: func(_ configtelemetry.Level) metric.MeterProvider {
+				return noop.NewMeterProvider()
+			},
+		},
+		handler,
+	)
+	if err != nil {
+		return nil, err
+	}
+	server := &httpServer{
+		Server: s,
+	}
+	if queryOpts.HTTP.TLSSetting != nil {
+		tlsCfg, err := queryOpts.HTTP.TLSSetting.LoadTLSConfig(ctx) // This checks if the certificates are correctly provided
+		if err != nil {
+			return nil, err
+		}
+		server.TLSConfig = tlsCfg
+	}
```

Comment:
@yurishkuro the `ToListener` function does the following 
```
func (hss *ServerConfig) ToListener(ctx context.Context) (net.Listener, error) {
	listener, err := net.Listen("tcp", hss.Endpoint)
	if err != nil {
		return nil, err
	}

	if hss.TLSSetting != nil {
		var tlsCfg *tls.Config
		tlsCfg, err = hss.TLSSetting.LoadTLSConfig(ctx)
		if err != nil {
			return nil, err
		}
		tlsCfg.NextProtos = []string{http2.NextProtoTLS, "http/1.1"}
		listener = tls.NewListener(listener, tlsCfg)
	}

	return listener, nil
}
```
However, this does not set the `TLSConfig` on the server itself. Does that mean we need to leave this bit in here? or is there a way to leverage OTEL's helper for this.

Reply:
I assume if it's on the listener already it does not need to be on the server. Do we not have unit tests with certificates? They should catch a problem if TLS is not initialized.

Reply:
@yurishkuro we do - but they fail if i remove the setting of `TLSConfig` from the server (see below)
```
--- FAIL: TestServerHTTPTLS (14.05s)
    --- FAIL: TestServerHTTPTLS/should_pass_with_TLS_client_to_trusted_TLS_server_with_correct_hostname (2.00s)
        /Users/mzaryab/Documents/opensource/jaeger/cmd/query/app/logger.go:146: 2024-10-25T18:30:13.747-0400	WARN	Using the 0.0.0.0 address exposes this server to every network interface, which may facilitate Denial of Service attacks.	{"documentation": "https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#safeguards-against-denial-of-service-attacks"}
        /Users/mzaryab/Documents/opensource/jaeger/cmd/query/app/logger.go:146: 2024-10-25T18:30:13.748-0400	WARN	Using the 0.0.0.0 address exposes this server to every network interface, which may facilitate Denial of Service attacks.	{"documentation": "https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#safeguards-against-denial-of-service-attacks"}
        /Users/mzaryab/Documents/opensource/jaeger/cmd/query/app/logger.go:146: 2024-10-25T18:30:13.749-0400	INFO	Using UI configuration	{"path": ""}
        /Users/mzaryab/Documents/opensource/jaeger/cmd/query/app/logger.go:146: 2024-10-25T18:30:13.749-0400	INFO	Query server started	{"http_addr": "[::]:53056", "grpc_addr": "[::]:53055"}
        /Users/mzaryab/Documents/opensource/jaeger/cmd/query/app/logger.go:146: 2024-10-25T18:30:13.749-0400	INFO	Starting HTTP server	{"port": 53056, "addr": ":0"}
        /Users/mzaryab/Documents/opensource/jaeger/cmd/query/app/logger.go:146: 2024-10-25T18:30:13.749-0400	INFO	Starting GRPC server	{"port": 53055, "addr": ":0"}
        /Users/mzaryab/Documents/opensource/jaeger/cmd/query/app/logger.go:146: 2024-10-25T18:30:13.749-0400	ERROR	Could not start HTTP server	{"error": "open : no such file or directory"}
        /Users/mzaryab/Documents/opensource/jaeger/cmd/query/app/server_test.go:432: 
            	Error Trace:	/Users/mzaryab/Documents/opensource/jaeger/cmd/query/app/server_test.go:432
            	Error:      	Received unexpected error:
            	            	context deadline exceeded
```

Reply:
ok, then can we set tls on the server too?

Reply:
yeah if i leave the highlighted block as is then its fine - just wanted to check if there was something we could to remove this however